src/foam/nanos/auth/CapabilityAuthService.js

```diff
@@ -17,20 +17,83 @@ foam.CLASS({
   ],
 
   javaImports: [
+    'foam.dao.Sink',
+    'foam.dao.ProxySink',
+    'foam.dao.LimitedSink',
     'foam.dao.ArraySink',
     'foam.dao.DAO',
+    'foam.core.Detachable',
+    'foam.core.X',
     'foam.mlang.predicate.Predicate',
     'foam.nanos.crunch.Capability',
     'foam.nanos.crunch.CapabilityJunctionStatus',
     'foam.nanos.crunch.UserCapabilityJunction',
     'foam.nanos.logger.Logger',
     'foam.nanos.session.Session',
+    'java.util.concurrent.ConcurrentHashMap',
     'java.util.Date',
     'java.util.List',
+    'java.util.Map',
     'static foam.mlang.MLang.*'
   ],
 
+  properties: [
+    {
+      class: 'Map',
+      name: 'cache',
+      javaFactory: `
+        return new ConcurrentHashMap<String, Boolean>();
+      `
+    },
+    {
+      name: 'initialized',
+      class: 'Boolean',
+      value: false
+    }
+  ],
+
   methods: [
+    {
+      name: 'initialize',
+      synchronized: true,
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+      ],
+      javaCode: `
+        if ( getInitialized() )
+          return;
+
+        DAO userCapabilityJunctionDAO = (DAO) x.get("userCapabilityJunctionDAO");
+        DAO capabilityDAO = (x.get("capabilityDAO") == null ) ? (DAO) x.get("capabilityDAO") : (DAO) x.get("localCapabilityDAO");
+        if ( capabilityDAO == null || userCapabilityJunctionDAO == null )
+          return;
```

Comment:
Should we still mark it as initialized?

Reply:
Would it be easier to just put this in the Object's init() method, which gets called when it is constructred?

Reply:
re: mark as initialized 
I think if either capabilityDAO or userCapabilityJunctionDAO is null, this shouldn't be initialized. if the daos are null then we cant attach the listeners on them in the initialization code, so it wouldneed to be re-initialized later on anyway

re: init
I tried to put this in the init method, but I think this is causing recursion and npe problems in boot since the serviceDAO_ used to create the context during boot uses the permissionedPropertyDAO which uses auth chain